internal/webhooks/cluster.go

```diff
@@ -56,9 +60,15 @@ func (webhook *Cluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
 // +kubebuilder:webhook:verbs=create;update;delete,path=/validate-cluster-x-k8s-io-v1beta1-cluster,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=clusters,versions=v1beta1,name=validation.cluster.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 // +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1beta1-cluster,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=clusters,versions=v1beta1,name=default.cluster.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
 
+// ClusterCacheTrackerReader is a scoped-down interface from ClusterCacheTracker that only allows to get a reader client.
+type ClusterCacheTrackerReader interface {
```

Comment:
What about moving this interface in "sigs.k8s.io/cluster-api/controllers/remote" (where ClusterCacheTracker is defined) so we can re-use it for other webhooks in future?

Reply:
I'll try but afaik we can't because we may have a circular dependency then.

Reply:
Yes, this is not possible because otherwise:

* `sigs.k8s.io/cluster-api/internal/webhooks` would have the new import `sigs.k8s.io/cluster-api/controllers/remote`

This would introduce a cyclic dependency because:

* `sigs.k8s.io/cluster-api/controllers/remote` imports `sigs.k8s.io/cluster-api/internal/test/envtest` (in tests)
* `sigs.k8s.io/cluster-api/internal/test/envtest` imports `sigs.k8s.io/cluster-api/internal/webhooks`

Adding the interface here also follows the best practice:

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values.

[[0]](https://go.dev/wiki/CodeReviewComments)